refactor: rename base_word to lemma

cblanken opened the pull request:
This addresses issue https://github.com/simjanos-dev/LinguaCafe/issues/99.

I was mainly looking at this, because I was planning to work a bit on issue https://github.com/simjanos-dev/LinguaCafe/issues/43 to figure out how word level tracking by lemma might work, and it seemed like a good idea to get this refactor out of the way before making any other changes.

A couple notes
1. at the moment this just does a refactor for references to base_word and base_word_reading. There are still some references to "baseWord" and "setBaseWord" functions in a few Vue components. @simjanos-dev would you like me to refactor those as well for consistency?
2. I wrote the migration for the `encountered_words` table to manually copy over each `base_word` to the `lemma` field. I thought it was prudent since there isn't really a guarantee that they're equivalent, but it may make the migration pretty lengthy for a database with a lot of encountered terms.


I tested the migration and rollback on a small test instance and did some rudimentary tests for my regular text import and reading workflow and everything seems to work fine, but I've not done anything exhaustive just yet.

cblanken committed:
refactor: rename base_word to lemma

Per issue #99, references to "base_word" in the EncounteredWords (encountered_words)
model should be "lemma" instead.

simjanos-dev commented:
Hi!

Thank you for the PR!

> There are still some references to "baseWord" and "setBaseWord" functions in a few Vue components. @simjanos-dev would you like me to refactor those as well for consistency?

Yes please. The expression `baseWord` should be replaced with `lemma` in every form. 

Which word should we use on the UI and manual?

> I thought it was prudent since there isn't really a guarantee that they're equivalent, but it may make the migration pretty lengthy for a database with a lot of encountered terms.

Damn, my code is terrible at some places. I think I literally wrote this part in the first 2 weeks when I started writing lc for myself. I think it is faster to delete the lemma column, then rename the base_word column to lemma. 

I also found some parts where both are present, when encountered_words are created.
```
$encounteredWord->lemma = '';
$encounteredWord->base_word = '';
```

I'll double check the changes when the PR is ready.

I'm currently working on refactoring controllers and services. Is it okay if I keep going, it may cause some conflicts?

cblanken commented:
> Yes please. The expression baseWord should be replaced with lemma in every form.

Will do. :saluting_face: 

> Which word should we use on the UI and manual?

I'd say "lemma" since it's already in use in the dictionary lookup view and is more accurate anyway. It might be a good idea to provide a couple examples or link to the definition in the manual though since the term's a bit esoteric for anyone not into linguistics.

> I think it is faster to delete the lemma column, then rename the base_word column to lemma.

Good call. I'll fixup the migration to do that instead.

> I'm currently working on refactoring controllers and services. Is it okay if I keep going, it may cause some conflicts?

Please go ahead. I'll rebase and resolve any conflicts if they come up before pushing anything else to this PR.

simjanos-dev commented:
Thank you for working on this!

cblanken committed:
refactor: improve migration performance

Records from the original `base_word` column don't need to be manually copied
to `lemma` since the `base_word` column will be replaced anyway. Instead, drop
the currently unused `lemma` column and then rename `base_word` -> `lemma`.

cblanken committed:
refactor: fix frontend references to baseWord

Some Vue components use the `baseWord` naming which should now refer to `lemma`
for consistency with the backend.

cblanken commented:
Okay, I've tried it on a backup of my own database and it seems to work fine, I do have one question though. At the moment I've left the `rollback` function to still copy back all the data from the `lemma`, but that takes quite a while (~10 minutes) on my machine.

Would you prefer that I simply rename the column again? To return to the original state during a rollback, the data should be copied back into `base_word` so that both the `base_word` and `lemma` columns contain the same values, but as far as I can tell, the project doesn't ever use `.lemma` so I don't think it would break anything if the `lemma` column was just renamed to `base_word` and then a new empty `lemma` column were created during a rollback. That would be faster, but all the `lemma` values would be empty.

simjanos-dev commented:
Yes, that would work.

cblanken committed:
refactor: improve rollback performance

Because, the `lemma` column is not referenced in the application until this refactor,
on rollback, the `lemma` column can simply be renamed to the original `base_word` name
and `lemma` can be created as an empty column.

simjanos-dev commented:
Thank you so much for this PR!
